### DISCUSSION TITLE
Notifications to users

### DISCUSSION THREAD

**Discussion by vinstinct** [laravel/framework#50175] [General]
Hello, I am trying to build a appointment booking app 

I have a function in my controller called cancelbooking function which will change the status of the booking from 0 to 1 will change that from from active bookings to cancel also will delete the related bookings in checkin table and also put back the canceled time back to available appointments by changing the status from 0 to 1. Now I Want to send a notification to the users that the appointment got canceled 

Here is the code 
use ``` php 

    public function cancelBooking(Request $request, string $id)
    {
        try {
            // Log the ID to verify if it's being received correctly
            info('Booking ID: ' . $id);

            // Find the booking by ID or throw an exception if not found
            $booking = Booking::findOrFail($id);

            // Log information about the booking
            info('Booking found: ' . $booking->id);

            // Update the booking attributes
            $booking->update([
                'status' => 1,
            ]);

            // Log a message after updating the booking
            info('Booking status updated: ' . $booking->status);

            // Find the related checkinout record and delete it
            $checkinout = Checkinout::where('bookingID', $booking->id)->first();
            if ($checkinout) {
                $checkinout->delete();
                // Log a message after deleting the checkinout record
                info('Checkinout record deleted for booking: ' . $booking->id);
            }

            // Update the status in the times table to make the appointment available again
            Time::where('id', $booking->appointment_id)->update(['status' => 0]);

            // Log a message after updating the status in the times table
            info('Status updated in the times table for appointment: ' . $booking->appointment_id);

            // Send notification to all users
            $users = User::all();
            Notification::send(User::all(), new BookingCanceled($booking->id));

            // Log a message after sending the notification
            info('Notification sent to all users');

            return redirect()->route('home.carrier.index')->with('message', 'Appointment canceled. Please go to the dashboard to see your appointments');
        } catch (\Exception $e) {
            // Log any exceptions that occur
            info('Exception caught: ' . $e->getMessage());

            // Validation failed or booking not found, store the errors in the session
            return redirect()->route('home.carrier.index')->withErrors(['error' => 'Validation failed or Booking not found. Please check your inputs and try again.']);
        }
    }

`<?php

namespace App\Notifications;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Notifications\Messages\MailMessage;
use Illuminate\Notifications\Notification;
use Illuminate\Queue\SerializesModels;

class BookingCanceled extends Notification
{
    use Queueable, SerializesModels;

    private $booking;

    /**
     * Create a new notification instance.
     *
     * @param mixed $booking
     */
    public function __construct($id)
    {
        $this->booking = $booking;
    }

    /**
     * Get the notification's delivery channels.
     *
     * @return array<int, string>
     */
    public function via($notifiable)
    {
        return ['database'];
    }

    /**
     * Get the mail representation of the notification.
     */
    public function toMail($notifiable)
    {
        return (new MailMessage)
                    ->line('The booking with ID ' . $this->booking->id . ' has been canceled.')
                    ->action('View Booking', route('booking.show', $this->booking->id))
                    ->line('Thank you for using our application!');
    }

    /**
     * Get the array representation of the notification.
     *
     * @return array<string, mixed>
     */
    public function toArray(object $notifiable): array
    {
        return [
            'booking_id' => $this->booking->id,
            'message' => 'The booking has been canceled.',
        ];
    }

    public function toDatabase($notifiable)
    {
        return [
            'booking_id' => $this->booking->id,
            'message' => 'The booking has been canceled.',
            'booking_status' => $this->booking->status,
            // Add other data you want to store in the notification
        ];
    }
}
`
When I checked the logs then this is what I see. I do not know why $bookings is undefined 

[2024-02-21 04:21:23] local.INFO: Booking ID: 141  
[2024-02-21 04:21:23] local.INFO: Booking found: 141  
[2024-02-21 04:21:23] local.INFO: Booking status updated: 1  
[2024-02-21 04:21:24] local.INFO: Checkinout record deleted for booking: 141  
[2024-02-21 04:21:24] local.INFO: Status updated in the times table for appointment: 350  
[2024-02-21 04:21:24] local.INFO: Exception caught: Undefined variable $booking

**Comment by henzeb** [comment]
could you fix the formatting of the BookingCanceled thing? also add the prefix php (use \`\`\` php ) on both. then we can read the code...

**Reply by vinstinct** [reply]
hello Henzeb, I have updated the format, let me know if you can see it and have any recommendations. Thanks

**Reply by henzeb** [reply]
Try setting the visibility of the booking property on your notification to protected. When queueing, it gets serialized, and then it can't see the private properties.
